### PR TITLE
feat(medusa): add analytics config

### PR DESCRIPTION
**Pull request by kasperkristensen**
**What**
- Adds new entity AnalyticsConfig
- Adds new service AnalyticsConfigService
- Adds new repository AnalyticsConfigRepository
- Adds new endpoints to get, create, update, and delete analytics configs

**Why**
As we begin gathering usage insights to help us improve Medusa, we want to give each individual users the ability to control what data they share with us, or not share any data with us at all. The AnalyticsConfig holds information that is used to check if the user wishes for their data to be anonymized or if they have opted out of sharing usage data.

The entire feature can be disabled on a store level by setting the feature flag `MEDUSA_FF_ANALYTICS=false` in their environment variables, the feature is enabled by default. 

**Testing**
Adds integration test for each of the new endpoints

Resolves CORE-656, CORE-655, CORE-654

Also resolves CORE-574

### CODE REVIEW

File: packages/medusa/src/services/analytics-config.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        MedusaError.Types.NOT_FOUND,
+        `No analytics config found for user with id: ${userId}`
+      )
+    }
+
+    return analyticsConfig
+  }
+
+  async create(userId: string, data: CreateAnalyticsConfig) {
+    return await this.atomicPhase_(async (transactionManager) => {
```
Comment (srindom) [inline packages/medusa/src/services/analytics-config.ts:54]: nit: atomic phase is not necessary here just use 

`const manager = this.transactionManager_ || this.manager_`

File: packages/medusa/src/services/analytics-config.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      for (const [key, value] of Object.entries(update)) {
+        config[key] = value
+      }
+
+      return await analyticsRepo.save(config)
+    })
+  }
+
+  async delete(userId: string) {
+    return await this.atomicPhase_(async (transactionManager) => {
```
Comment (srindom) [inline packages/medusa/src/services/analytics-config.ts:85]: nit: atomic phase is not necessary here just use 

`const manager = this.transactionManager_ || this.manager_`

**Review by srindom — COMMENTED** [review]
Generally looks great - have a few minor comments :)

File: packages/medusa/src/types/analytics-config.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+export type CreateAnalyticsConfig = {
+  opt_out: boolean
+  anonymize: boolean
+}
+
+export type UpdateAnalyticsConfig = {
+  opt_out: boolean
+  anonymize: boolean
```
Comment (pKorsholm) [inline packages/medusa/src/types/analytics-config.ts:8]: question: shouldn't these be optional since it's an update, and in the request to update the properties are optional?

File: packages/medusa/src/services/analytics-config.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+  async update(userId: string, update: UpdateAnalyticsConfig) {
+    return await this.atomicPhase_(async (transactionManager) => {
+      const analyticsRepo = transactionManager.getCustomRepository(
+        this.analyticsConfigRepository_
+      )
+
+      const config = await this.retrieve(userId)
+
+      for (const [key, value] of Object.entries(update)) {
```
Comment (pKorsholm) [inline packages/medusa/src/services/analytics-config.ts:76]: if properties are optional a check for `typeof value !== 'undefined'`

File: packages/medusa/src/services/user.ts [hunk 5]
```diff
@@ -244,6 +262,14 @@ class UserService extends TransactionBaseService {
         return Promise.resolve()
       }
 
+      if (this.featureFlagRouter_.isFeatureEnabled(AnalyticsFeatureFlag.key)) {
+        try {
+          await analyticsServiceTx.delete(userId)
+        } catch (error) {
+          console.warn(error)
+        }
```
Comment (pKorsholm) [inline packages/medusa/src/services/user.ts:270]: Question: this shouldn't be needed if the delete is idempotent, is there a reason for the warning here (such as missing migrations etc)?

File: packages/medusa/src/api/routes/admin/analytics-configs/create-analytics-config.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
+import { IsBoolean } from "class-validator"
+import { Request, Response } from "express"
+import { EntityManager } from "typeorm"
+import { AnalyticsConfigService } from "../../../../services"
+import { CreateAnalyticsConfig } from "../../../../types/analytics-config"
+
+export default async (req: Request, res: Response) => {
```
Comment (pKorsholm) [inline packages/medusa/src/api/routes/admin/analytics-configs/create-analytics-config.ts:7]: question: should we add a comment about the missing oas here so it's not accidentally added later on? (same for all endpoints)

File: integration-tests/api/__tests__/admin/analytics-config.js [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+
+    medusaProcess.kill()
+  })
+
+  describe("GET /admin/analytics-config", () => {
+    beforeEach(async () => {
+      try {
+        await adminSeeder(dbConnection)
+      } catch (e) {
+        console.error(e)
```
Comment (pKorsholm) [inline integration-tests/api/__tests__/admin/analytics-config.js:44]: rm try/catch

File: integration-tests/api/__tests__/admin/analytics-config.js [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+          opt_out: false,
+          anonymize: false,
+          created_at: expect.any(String),
+          updated_at: expect.any(String),
+        },
+      })
+    })
+
+    it("should return 404 if no config exists", async () => {
+      const api = useApi()
```
Comment (pKorsholm) [inline integration-tests/api/__tests__/admin/analytics-config.js:73]: Add expect assertions since no assertions is a success test
```suggestion
      expect.assertions(2)
      const api = useApi()
```

File: integration-tests/api/__tests__/admin/analytics-config.js [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+    })
+  })
+
+  describe("POST /admin/analytics-config", () => {
+    beforeEach(async () => {
+      try {
+        await adminSeeder(dbConnection)
+      } catch (e) {
+        console.error(e)
+      }
```
Comment (pKorsholm) [inline integration-tests/api/__tests__/admin/analytics-config.js:91]: rm try/catch

File: integration-tests/api/__tests__/admin/analytics-config.js [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+    })
+  })
+
+  describe("POST /admin/analytics-config/update", () => {
+    beforeEach(async () => {
+      try {
+        await adminSeeder(dbConnection)
+      } catch (e) {
+        console.error(e)
+      }
```
Comment (pKorsholm) [inline integration-tests/api/__tests__/admin/analytics-config.js:131]: rm try/catch

File: integration-tests/api/__tests__/admin/analytics-config.js [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+          opt_out: true,
+          anonymize: false,
+          created_at: expect.any(String),
+          updated_at: expect.any(String),
+        },
+      })
+    })
+
+    it("should fail update the config of the logged in user if a config does not exist", async () => {
+      const api = useApi()
```
Comment (pKorsholm) [inline integration-tests/api/__tests__/admin/analytics-config.js:165]: add expect assertions

File: integration-tests/api/__tests__/admin/analytics-config.js [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+    })
+  })
+
+  describe("DELETE /admin/analytics-config", () => {
+    beforeEach(async () => {
+      try {
+        await adminSeeder(dbConnection)
+      } catch (e) {
+        console.error(e)
+      }
```
Comment (pKorsholm) [inline integration-tests/api/__tests__/admin/analytics-config.js:189]: rm try/catch

File: integration-tests/api/__tests__/admin/user.js [hunk 13]
```diff
@@ -17,7 +22,7 @@ describe("/admin/users", () => {
   beforeAll(async () => {
     const cwd = path.resolve(path.join(__dirname, "..", ".."))
     dbConnection = await initDb({ cwd })
-    medusaProcess = await setupServer({ cwd })
+    medusaProcess = await setupServer({ cwd, verbose: true })
```
Comment (pKorsholm) [inline integration-tests/api/__tests__/admin/user.js:25]:
```suggestion
    medusaProcess = await setupServer({ cwd, verbose: false })
```

File: integration-tests/api/__tests__/admin/user.js [hunk 14]
```diff
@@ -322,11 +327,9 @@ describe("/admin/users", () => {
 
       const usersBeforeDelete = usersBeforeDeleteResponse.data.users
 
-      const response = await api
-        .delete(`/admin/users/${userId}`, {
-          headers: { Authorization: "Bearer test_token" },
-        })
-        .catch((err) => console.log(err))
+      const response = await api.delete(`/admin/users/${userId}`, {
+        headers: { Authorization: "Bearer test_token" },
```
Comment (pKorsholm) [inline integration-tests/api/__tests__/admin/user.js:331]: nit: refactor headers to a const like in `analytics-config`

**Review by pKorsholm — COMMENTED** [review]
Very nice! generally it looks really good, I have a couple of minor changes for the integration tests and a small comment about updating.

Also agree with @srindom that the default value probably should be false for the feature flag before the feature is complete 😄

File: packages/medusa/src/loaders/feature-flags/analytics.ts [hunk 15]
```diff
@@ -0,0 +1,11 @@
+import { FlagSettings } from "../../types/feature-flags"
+
+const AnalyticsFeatureFlag: FlagSettings = {
+  key: "analytics",
+  default_val: true,
```
[the remark this answers is not in the document]
Comment (kasperkristensen) [reply packages/medusa/src/loaders/feature-flags/analytics.ts:5]: Sure, though I thought the idea was that we wanted to get started with gathering analytics immediedtly without users needing to do anything actively to participate?

[the remark this answers is not in the document]
Comment (srindom) [reply packages/medusa/src/loaders/feature-flags/analytics.ts]: yes, but just until we have everything done with admin etc. so if we need to do additional work we haven't switched on the feature flag until the entire feature is ready - does that make sense?

File: integration-tests/api/__tests__/admin/analytics-config.js (hunk 8 above)
Comment (adrien2p) [reply integration-tests/api/__tests__/admin/analytics-config.js:69]: don't think the expect assertions is necessary here. it just means that we expect 3 expects to be called before the tests timeout. It seam a bit redundant no?

File: integration-tests/api/__tests__/admin/analytics-config.js [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+          user_id: "admin_user",
+          opt_out: true,
+          anonymize: false,
+          created_at: expect.any(String),
+          updated_at: expect.any(String),
+        },
+      })
+    })
+
+    it("should fail update the config of the logged in user if a config does not exist", async () => {
```
Comment (adrien2p) [inline integration-tests/api/__tests__/admin/analytics-config.js:154]: suggestion(non blocking): Maybe we could create the config if it does not exists yet. since it is link to the logged in user it could make sense, and also would make the update idempotent. wdyt?

File: integration-tests/api/__tests__/admin/analytics-config.js (hunk 11 above)
Comment (adrien2p) [reply integration-tests/api/__tests__/admin/analytics-config.js:155]: suggestion: We should use .catch instead of try catch block in order to always reach the expect. and therefore the expect assertions become redundant

File: integration-tests/api/__tests__/admin/user.js [hunk 17]
```diff
@@ -354,3 +357,74 @@ describe("/admin/users", () => {
...
+      const api = useApi()
+
+      const userId = "member-user"
+
+      await simpleAnalyticsConfigFactory(dbConnection, {
+        user_id: userId,
+      })
+
+      const response = await api.delete(`/admin/users/${userId}`, {
+        headers: { Authorization: "Bearer test_token" },
```
Comment (adrien2p) [inline integration-tests/api/__tests__/admin/user.js:403]: suggestion: as mention by phil, the headers can be moved to a const to be used accros the test cases

File: packages/medusa/src/api/routes/admin/analytics-configs/index.ts [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+
+  route.delete(
+    "/",
+    middlewares.wrap(require("./delete-analytics-config").default)
+  )
+
+  return app
+}
+
+export type AdminAnalyticsConfigRes = {
```
Comment (adrien2p) [inline packages/medusa/src/api/routes/admin/analytics-configs/index.ts:35]: todo: we are missing the delete res type here.

File: packages/medusa/src/migrations/1665660449659-add_analytics_config.ts [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+import AnalyticsFeatureFlag from "../loaders/feature-flags/analytics"
+
+export const featureFlag = AnalyticsFeatureFlag.key
+
+export class addAnalyticsConfig1665660449659 implements MigrationInterface {
+  name = "addAnalyticsConfig1665660449659"
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
+      `CREATE TABLE "analytics_config" ("id" character varying NOT NULL, "created_at" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(), "updated_at" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(), "deleted_at" TIMESTAMP WITH TIME ZONE, "user_id" character varying NOT NULL, "opt_out" boolean NOT NULL DEFAULT false, "anonymize" boolean NOT NULL DEFAULT false, CONSTRAINT "PK_ab46f7f4004ed899d568982c06c" PRIMARY KEY ("id", "user_id"))`
```
Comment (adrien2p) [inline packages/medusa/src/migrations/1665660449659-add_analytics_config.ts:11]: todo: The primary key here will enforce uniqueness even if the config is deleted as the tuple still exists.
instead, I think that you should create a unique index id, user_id where deleted_at is null

File: packages/medusa/src/services/analytics-config.ts [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+
+  constructor({ analyticsConfigRepository, manager }: InjectedDependencies) {
+    // eslint-disable-next-line prefer-rest-params
+    super(arguments[0])
+
+    this.manager_ = manager
+    this.analyticsConfigRepository_ = analyticsConfigRepository
+  }
+
+  async retrieve(userId: string) {
```
Comment (adrien2p) [inline packages/medusa/src/services/analytics-config.ts:32]: todo: Missing return type

File: packages/medusa/src/services/analytics-config.ts [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+      throw new MedusaError(
+        MedusaError.Types.NOT_FOUND,
+        `No analytics config found for user with id: ${userId}`
+      )
+    }
+
+    return analyticsConfig
+  }
+
+  async create(userId: string, data: CreateAnalyticsConfig) {
```
Comment (adrien2p) [inline packages/medusa/src/services/analytics-config.ts:53]: todo: Missing return type

Comment (adrien2p) [reply packages/medusa/src/services/analytics-config.ts:53]: todo: Add an atomicPhase for mutation method

File: packages/medusa/src/services/analytics-config.ts [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+
+    try {
+      const config = analyticsRepo.create({ user_id: userId, ...data })
+      return await analyticsRepo.save(config)
+    } catch (error) {
+      throw formatException(error)
+    }
+  }
+
+  async update(userId: string, update: UpdateAnalyticsConfig) {
```
Comment (adrien2p) [inline packages/medusa/src/services/analytics-config.ts:67]: todo: Missing return type and atomicPhase

File: packages/medusa/src/services/analytics-config.ts [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+    const manager = this.transactionManager_ || this.manager_
+    const analyticsRepo = manager.getCustomRepository(
+      this.analyticsConfigRepository_
+    )
+
+    try {
+      const config = analyticsRepo.create({ user_id: userId, ...data })
+      return await analyticsRepo.save(config)
+    } catch (error) {
+      throw formatException(error)
```
Comment (adrien2p) [inline packages/medusa/src/services/analytics-config.ts:63]: suggestion(non blocking): I think that the formatException should be moved to the error-handler middleware to be used accros the entire code base error. Therefore it could be removed from here and in another pr it could be removed from the whole code base. wdyt?

File: packages/medusa/src/services/analytics-config.ts (hunk 22 above)
Comment (kasperkristensen) [reply packages/medusa/src/services/analytics-config.ts:67]: Removed atomicPhase as Seb mentioned it was not necessary in this case

Comment (adrien2p) [reply packages/medusa/src/services/analytics-config.ts]: alright, got it, I suppose we don't expect a user to consume it ahah

File: packages/medusa/src/migrations/1665660449659-add_analytics_config.ts (hunk 19 above)
Comment (kasperkristensen) [reply packages/medusa/src/migrations/1665660449659-add_analytics_config.ts:11]: Something along the lines of this?

```suggestion
      await queryRunner.query(`CREATE TABLE "analytics_config" ("id" character varying NOT NULL, "created_at" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(), "updated_at" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(), "deleted_at" TIMESTAMP WITH TIME ZONE, "user_id" character varying NOT NULL, "opt_out" boolean NOT NULL DEFAULT false, "anonymize" boolean NOT NULL DEFAULT false, CONSTRAINT "PK_93505647c5d7cb479becb810b0f" PRIMARY KEY ("id"))`);
        await queryRunner.query(`CREATE UNIQUE INDEX "IDX_379ca70338ce9991f3affdeedf" ON "analytics_config" ("id", "user_id") WHERE deleted_at IS NULL`);
```

File: integration-tests/api/__tests__/admin/analytics-config.js (hunk 16 above)
Comment (kasperkristensen) [reply integration-tests/api/__tests__/admin/analytics-config.js:154]: We can definitely do that, it's not something that would ever occur, but I'll add it if we ever switch up how we handle adding and updating these from the admin 😄

File: packages/medusa/src/migrations/1665660449659-add_analytics_config.ts (hunk 19 above)
Comment (adrien2p) [reply packages/medusa/src/migrations/1665660449659-add_analytics_config.ts]: yes exactly, I prefer to use human readable indeces names but the choice is yours. usually I compose the name like follow {constraint_type}_{table_name}_{columns}. but otherwise it is very nice ahah

**Review by srindom — APPROVED** [review]
LGTM!

File: integration-tests/api/__tests__/admin/analytics-config.js (hunk 8 above)
Comment (pKorsholm) [reply integration-tests/api/__tests__/admin/analytics-config.js:69]: not neccessary, this comment was made before the test was refactored at which point the assertions was made in the catch block 😄 closing the comment
